compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        }
+                    }
+                    None => {
+                        // We didn't find any issues, so we need to push the algorithm forward
+                        // First, eliminate any arguments that currently satisfy their inputs
+                        let mut i = cmp::min(arg_indexes.len(), input_indexes.len());
+                        while i > 0 {
+                            let idx = i - 1;
+                            if compatibility_matrix[idx][idx] {
+                                satisfy_input(
+                                    &mut compatibility_matrix,
+                                    &mut input_indexes,
+                                    &mut arg_indexes,
+                                    idx,
+                                    idx,
+                                );
+                            }
+                            i -= 1;
+                        }
+                    }
+                };
+            }
+
+            let issue_count = issues.len();
+            if issue_count > 0 {
+                // We found issues, so lets construct a diagnostic that summarizes the issues we found
+                // FIXME: This might need some refining in code review
+                let mut labels = vec![];
+                let mut suggestions = vec![];
+                let source_map = self.sess().source_map();
+                for issue in issues {
+                    match issue {
+                        Issue::Invalid(arg) => {
+                            let span = provided_args[arg].span;
+                            let expected_type = expected_input_tys[arg];
+                            let found_type = final_arg_types.iter().filter(|(i, _, _)| *i == arg).next().unwrap().1;
+                            labels.push((span, format!("expected {}, found {}", expected_type, found_type)));
+                        }
+                        Issue::Extra(arg) => {
+                            // FIXME: This could probably be a lot cleaner, but I dunno how
+                            let span = provided_args[arg].span;
+                            let hungry_span = if arg < provided_args.len() - 1 {
+                                // Eat the comma
+                                Span::new(
+                                    span.lo(),
+                                    span.hi() + BytePos(2),
+                                    span.ctxt(),
+                                )
+                            } else {
+                                span
+                            };
+                            let found_type = self.check_expr(&provided_args[arg]);
+                            let fn_name = fn_def_id
+                                .and_then(|def_id| tcx.hir().get_if_local(def_id))
+                                .and_then(|node| node.ident())
+                                .map(|ident| format!("fn {}(..)", ident))
```

Comment:
Maybe use `def_kind.descr()` here, like you do on line 742 below?

Reply:
this seems to only return `function`, rather than the name of the function.  Unless you just mean before the ident:

```
no parameter of type {} is needed for function complex
```

Reply:
That's not what I meant but it's better than what I had in mind :laughing: so yes, I think that would be a good change.